### PR TITLE
[rfr] Gnocchi Metrics SDK v1

### PR DESCRIPTION
**Pull request by haphan**
This is the proposal for Gnocchi Metric service SDK implementation. 

Official Documentation: http://gnocchi.xyz/stable_4.0/rest.html

@jamiehannaford Comments would be much appreciated!!

### CODE REVIEW

**Comment by jamiehannaford** [comment]
LGTM 👍  just need some tests

File: src/Metric/v1/Gnocchi/Models/Resource.php [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        'ended_at'              => 'endedAt',
+    ];
+
+    public function retrieve()
+    {
+        $response = $this->execute($this->api->getResource(), ['type' => $this->type, 'id' => $this->id]);
+        $this->populateFromResponse($response);
+    }
+
+
```
Comment (jamiehannaford) [inline src/Metric/v1/Gnocchi/Models/Resource.php]: nit: remove extra space

File: src/Metric/v1/Gnocchi/Params.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        return [
+            'location'    => self::HEADER,
+            'type'        => self::STRING_TYPE,
+            'sentAs'      => 'Content-Type',
+            'description' => 'Override request header',
+            'documented'  => false,
+        ];
+    }
+
+    public function idUrl($type)
```
Comment (jamiehannaford) [inline src/Metric/v1/Gnocchi/Params.php]: missing return type here

File: src/Metric/v1/Gnocchi/Service.php [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+        return $this->model(Resource::class)->enumerate($this->api->getResources(), $options);
+    }
+
+    /**
+     * @param array $options
+     *
+     * @return \OpenStack\Metric\v1\Gnocchi\Models\Resource
+     */
+    public function getResource(array $options = []): \OpenStack\Metric\v1\Gnocchi\Models\Resource
```
Comment (jamiehannaford) [inline src/Metric/v1/Gnocchi/Service.php]: Do we need this fully qualified name or just `Resource`?

**Comment by haphan** [comment]
@jamiehannaford This pr is ready for review. Thanks

File: tests/unit/Metric/v1/Gnocchi/Models/ResourceTest.php [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        $this->assertInstanceOf(Metric::class, $metric);
+        $this->assertEquals($metric->name, 'storage.objects.outgoing.bytes');
+        $this->assertEquals($metric->id, '000b7bf8-0271-46dd-90aa-cfe89026a55a');
+    }
+
+    public function test_it_gets_metric_measures()
+    {
+        $this->setupMock('GET', sprintf('v1/resource/generic/1111/metric/storage.objects.outgoing.bytes/measures'), [], [], 'resource-metric-measures-get');
+
+        $this->resource->getMetricMeasures(['metric' => 'storage.objects.outgoing.bytes']);
```
Comment (jamiehannaford) [inline tests/unit/Metric/v1/Gnocchi/Models/ResourceTest.php]: same here

File: tests/unit/Metric/v1/Gnocchi/Models/ResourceTest.php [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        $this->resource = new Resource($this->client->reveal(), new Api());
+        $this->resource->type = 'generic';
+        $this->resource->id = '1111';
+    }
+
+    public function test_it_retrieves()
+    {
+        $this->setupMock('GET', 'v1/resource/generic/1111', null, [], new Response(204));
+
+        $this->resource->retrieve();
```
Comment (jamiehannaford) [inline tests/unit/Metric/v1/Gnocchi/Models/ResourceTest.php]: can we do some assertions against the returned data?

File: tests/unit/Metric/v1/Gnocchi/Models/ResourceTest.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        $this->resource->getMetricMeasures(['metric' => 'storage.objects.outgoing.bytes']);
+    }
+
+    public function test_it_lists_resource_metrics()
+    {
+        $this->setupMock('GET', 'v1/resource/generic/1111/metric', [], [], 'resource-metrics-get');
+
+        $result = $this->resource->listResourceMetrics();
+
+        $this->assertContainsOnlyInstancesOf(Metric::class, $result);
```
Comment (jamiehannaford) [inline tests/unit/Metric/v1/Gnocchi/Models/ResourceTest.php]: maybe also check the count?

File: tests/unit/Metric/v1/Gnocchi/ServiceTest.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    public function test_it_lists_resource_types()
+    {
+        $this->client
+            ->request('GET', 'v1/resource_type', ['headers' => []])
+            ->shouldBeCalled()
+            ->willReturn($this->getFixture('resourcetypes-get'));
+
+        $result = $this->service->listResourceTypes();
+
+        $this->assertContainsOnlyInstancesOf(ResourceType::class, $result);
```
Comment (jamiehannaford) [inline tests/unit/Metric/v1/Gnocchi/ServiceTest.php]: check count

File: tests/unit/Metric/v1/Gnocchi/ServiceTest.php [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    public function test_it_lists_resources()
+    {
+        $this->client
+            ->request('GET', 'v1/resource/generic', ['headers' => [], 'query' => ['limit' => 3]])
+            ->shouldBeCalled()
+            ->willReturn($this->getFixture('resources-get'));
+
+        $result = $this->service->listResources(['limit' => 3]);
+
+        $this->assertContainsOnlyInstancesOf(Resource::class, $result);
```
Comment (jamiehannaford) [inline tests/unit/Metric/v1/Gnocchi/ServiceTest.php]: check count

File: tests/unit/Metric/v1/Gnocchi/ServiceTest.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    public function test_it_get_resource()
+    {
+        $this->assertInstanceOf(Resource::class, $this->service->getResource(['id' => '1']));
```
Comment (jamiehannaford) [inline tests/unit/Metric/v1/Gnocchi/ServiceTest.php]: check ID

**Comment by haphan** [comment]
@jamiehannaford Code review fixed.

**Comment by jamiehannaford** [comment]
Great job again @haphan 🎆

**Pull request merged**